webapi/src/test/scala/org/knora/webapi/store/triplestore/api/TriplestoreServiceFake.scala

```diff
@@ -0,0 +1,147 @@
+import org.knora.webapi.messages.util.rdf.QuadFormat
+import org.knora.webapi.messages.util.rdf.SparqlSelectResult
+import org.knora.webapi.messages.util.rdf.SparqlSelectResultBody
+import org.knora.webapi.messages.util.rdf.SparqlSelectResultHeader
+import org.knora.webapi.messages.util.rdf.VariableResultsRow
+
+final case class TriplestoreServiceFake(datasetRef: Ref[Dataset]) extends TriplestoreService {
+
+  override def doSimulateTimeout(): UIO[SparqlSelectResult] = ???
+
+  override def sparqlHttpSelect(
+    sparql: String,
+    simulateTimeout: Boolean,
+    isGravsearch: Boolean
+  ): UIO[SparqlSelectResult] = {
+    require(!simulateTimeout, "`simulateTimeout` parameter is not supported by fake implementation yet")
+    require(!isGravsearch, "`isGravsearch` parameter is not supported by fake implementation yet")
+
+    ZIO.scoped(execSelect(sparql).map(toSparqlSelectResult)).orDie
+  }
+
+  private def execSelect(query: String): ZIO[Any with Scope, Throwable, ResultSet] = {
+    def executeQuery(qExec: QueryExecution) = ZIO.attempt(qExec.execSelect)
+    def closeResultSet(rs: ResultSet)       = ZIO.succeed(rs.close())
+    getQueryExecution(query).flatMap(qExec => ZIO.acquireRelease(executeQuery(qExec))(closeResultSet))
+  }
+
+  private def getQueryExecution(query: String): ZIO[Any with Scope, Throwable, QueryExecution] = {
+    def acquire(query: String, ds: Dataset) = ZIO.attempt(QueryExecutionFactory.create(query, ds))
+    def release(qExec: QueryExecution)      = ZIO.succeed(qExec.close())
+    datasetRef.get.flatMap(ds => ZIO.acquireRelease(acquire(query, ds))(release))
+  }
+
+  private def toSparqlSelectResult(resultSet: ResultSet): SparqlSelectResult = {
+    val header     = SparqlSelectResultHeader(resultSet.getResultVars.asScala.toList)
+    val resultBody = getSelectResultBody(resultSet)
+    SparqlSelectResult(header, resultBody)
+  }
+
+  private def getSelectResultBody(resultSet: ResultSet): SparqlSelectResultBody = {
+    val rows: Seq[VariableResultsRow] = resultSet.asScala
+      .foldRight(List.empty[VariableResultsRow])((solution, list) => list.prepended(asVariableResultsRow(solution)))
+    SparqlSelectResultBody(rows)
+  }
+
+  private def asVariableResultsRow(solution: QuerySolution): VariableResultsRow = {
+    val keyValueMap = solution.varNames.asScala.map { key =>
+      val node = solution.get(key).asNode
+      val value: String = // do not include datatype in string if node is a literal
+        if (node.isLiteral) { node.getLiteralLexicalForm }
+        else { node.toString }
+      key -> value
+    }.toMap
+    VariableResultsRow(keyValueMap)
+  }
+
+  override def sparqlHttpAsk(query: String): UIO[SparqlAskResponse] =
+    ZIO.scoped(getQueryExecution(query).map(_.execAsk())).map(SparqlAskResponse).orDie
+
+  override def sparqlHttpConstruct(sparqlConstructRequest: SparqlConstructRequest): UIO[SparqlConstructResponse] = ???
```

Comment:
Are you going to impelment these methods later?

Reply:
Yes, we should do it. I have only implemented the ones which I needed for my tests.

I have been thinking about using this in the course of trying to speed up our integration-tests using an in-memory implementation.